src/mds/MDBalancer.cc

```diff
@@ -1154,7 +1154,10 @@ void MDBalancer::maybe_fragment(CDir *dir, bool hot)
       !dir->inode->is_stray()) { // not straydir
 
     // split
-    if (g_conf()->mds_bal_split_size > 0 && (dir->should_split() || hot)) {
+    if (dir->frag.bits() < dir->inode->get_expected_file_bits()) {
```

Comment:
I think it's better to move this logic into CDir::should_split().   Also need to update CDir::should_merge()

Reply:
That's where I had it originally, but after looking at the code both ways I ended up preferring having the logic centrally located in maybe_fragment().  I'm ok either way though.

Reply:
I think about this again. this is for burst operations on a directory. how about do the dirfrag splits on handling the setvxattr request.

Reply:
Hah!  I had the code there for a bit too. :smile: Lots of ways we could do it.  Do you have a preference?

Reply:
I prefer to do the split on handling the setvxattr request. because it's a hint, I prefer
to not persistent expected_files in inode_t.

Reply:
ok, so if we do not persist expected_files but expected_bits only, users won't be able to read back the expected_files later.  I'm ok with that as a trade-off if we think the inode_t is too big.  I think we could probably also switch expected_files_bits to a unit8_t so only a single byte increase in size.  I'll go ahead and give it a try.

Reply:
because it's a hint. I'm worry about that user may set and forget it.  but It keeps dir fragmented forever.

Reply:
Hrm, perhaps I misunderstand, but if we do not persist it, how would we prevent those directories from being merged in the future if they are empty?  Sort of the point of the option is for the user specifically to say "at some point in the future this directory is going to have millions of objects, so fragment it with that assumption".

Reply:
CDir has some field that track how hot it is. we can make resulting dirfrags very hot. which should prevent the from merging for a while

Reply:
If we were to take that approach, how would the user know if the hint to fragment is still being respected?  What I like about the first approach is that it is fairly deterministic.  If you set an expected value of X, CephFS will pre-create the fragments such that no fragmenting or merging will occur until X is exceeded.  If we reuse the hot/cold machinery, we are back to a situation where merging or fragmenting might happen behind the user's back without them knowing when or why.

Reply:
have you tried setting 'mds bal split bits' to larger value (also adjust split size and merge size), for example 8. With larger split bits, we can ensure dirfrag only get fragmented  once during the test.

Reply:
I did try changing mds_bal_split_bits and also tried adjusting the split size in different ways but seemed to generally make things worse.  I believe we still end up in sort of rough situations but with different pause duration and contention characteristics.  This isn't exactly right due to psuedo-random distribution of files across fragments, but should sort of be close to what happens I think:

```
Default: 

Time 000.0:    1 Fragment, 0 Files
Time 000.4:    8 Fragments, 10,000 Files
Time 003.2:   64 Fragments, 80,000 Files
Time 025.6:  512 Fragments, 640,000 Files
Time 204.8: 4096 Fragments, 5,120,000 Files

Total splits: 585
Files per split: 10,000
Files involved in all splits: 5,850,000
Resulting Fragments: 4096
```

By default we have more overall split events but the curve is shallower than if we set the split bits higher:

```
mds_bal_split_bits = 8
Time 000.0: 1 Fragment, 0 Files
Time 000.4: 256 Fragments, 10,000 Files
Time 010.2: 65536 Fragments, 256,000 Files

Total splits: 257
Files per split: 10,000
Files involved in all splits: 266,000
Resulting Fragments: 65536
```

With more split bits we can reduce the number of splits for this specific case, but we end up with way too many fragments.

```
expected_files = 15,000,000, mds_bal_split_bits = 3 // 15,000,000 < (2^11)*10,000
Time 000.0: 4096 Fragments, 0 Files // 2^(3 + 3 + 3 + 3) >= 2^11 = 4096

Total Splits: 585
Files per split: 0
Files involved in all splits: 0
Resulting Fragments: 4096
```

If we run for a shorter duration (say 20 seconds), the downside of setting a higher mds_bal_split_size is magnified:

```
Default: 

Time 000.0:    1 Fragment, 0 Files
Time 000.4:    8 Fragments, 10,000 Files
Time 003.2:   64 Fragments, 80,000 Files

Total splits: 9
Files per split: 10,000
Files involved in all splits: 90,000
Resulting Fragments: 64
```


```
mds_bal_split_bits = 8: 

Time 000.0: 1 Fragment, 0 Files
Time 000.4: 256 Fragments, 10,000 Files
Time 010.2: 65536 Fragments, 256,000 Files

Total splits: 257
Files per split: 10,000
Files involved in all splits: 266,000
Resulting Fragments: 65536
```

```
expected_files = 500,000, mds_bal_split_bits = 3 // 500,000 < (2^6)*10,000
Time 000.0: 4096 Fragments, 0 Files // 2^(3 + 3) >= 2^6 = 64

Total Splits: 9
Files per split: 0
Files involved in all splits: 0
Resulting Fragments: 64
```

Reply:
Make sense. I will do some local test for this PR

Reply:
Ok great!  Please let me know if you have other concerns as well.  I do wonder if we should store the expected_files as a uint32_t and compute the bits on the fly, so that the user can later view the xattr and see what expected_files is set to.  That would use 4bytes instead of 1byte in the inode.  I'm not sure how large the inodes are already, but they seem fairly large.  I don't know if the extra overhead would be a problem or not.

Reply:
I prefer not to store same information twice.

Reply:
> I prefer not to store same information twice.

Ok, what do you think about storing the expected_files in 4 bytes instead of bits in 1 byte?  The expected_files takes more space, but would allow users to check what the xattr is set to later on.  The bits takes less space, but we lose the value for what users set the files to.  We could also have users set the bits directly, but I think that users will not understand it very well (files is easier for them to understand).

What do you think?